cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -307,11 +398,52 @@ deprecated class UninitializedNode extends Node {
  * This class exists to match the interface used by Java. There are currently no non-abstract
  * classes that extend it. When we implement field flow, we can revisit this.
  */
-abstract class PostUpdateNode extends InstructionNode {
+abstract class PostUpdateNode extends Node {
   /**
    * Gets the node before the state update.
    */
   abstract Node getPreUpdateNode();
+
+  override Function getFunction() { result = getPreUpdateNode().getFunction() }
+
+  override IRType getType() { result = getPreUpdateNode().getType() }
+
+  override Location getLocation() { result = getPreUpdateNode().getLocation() }
+}
+
+/**
+ * A partial definition of a node. A partial definition that targets arrays or pointers is attached to
+ * an `InstructionNode` (specifially, to the `ChiInstruction` that follows the `StoreInstruction`), and
+ * a partial definition that targets a `FieldNode` is attached to the `FieldNode`.
+ *
+ * The pre-update node of a partial definition of a `FieldNode` is the `FieldNode` itself. This ensures
+ * that the data flow library's reverse read mechanism builds up the correct access path for nested
+ * fields.
+ * For instance, in `a.b.c = x` there is a partial definition for `c` (let's call it `post[c]`) and a
+ * partial definition for `b` (let's call it `post[b]`), and there is a read step from `b` to `c`
+ * (using `instrToFieldNodeReadStep`), so there is a store step from `post[c]` to `post[b]`.
+ */
+private newtype TPartialDefinition =
+  MkPartialDefinition(Node node) {
+    isPointerStoreNode(node, _, _) or
+    isArrayStoreNode(node, _, _) or
+    node instanceof FieldNode
+  }
+
+/** INTERNAL: do not use. A partial definition of a node. */
+abstract class PartialDefinition extends TPartialDefinition {
+  Node node;
+
+  PartialDefinition() { this = MkPartialDefinition(node) }
+
+  /** Gets the node before the state update. */
+  abstract Node getPreUpdateNode();
+
+  /** Gets the expression that is partially defined by this node. */
+  abstract Expr getDefinedExpr();
+
+  /** Gets a string representation of this partial definition. */
+  string toString() { result = node.toString() + " [partial definition]" }
```

Comment:
Does this ever get called?

Reply:
I don't think so. It has been useful for debugging but I'm totally okay with throwing it away.

Reply:
Removed in a00bd7ae027d17a2cd2e2931a37baa31943d1245.